### PR TITLE
feat(employee listing): Add Employee Listing Page with Dual Views and Advanced Functionalities for Enhanced User Experience

### PR DESCRIPTION
**Pull request by seraGeorge**
## Description

This pull request introduces an Employee Listing Page with dual views (List and Grid) and advanced functionalities to enhance the overall user experience. The key features include:

1. **Dual Views:**
   - Implemented both List and Grid views for flexible user preferences.

2. **Pagination and Infinite Loading:**
   - List view features pagination for organized navigation.
   - Grid view incorporates infinite loading for a seamless user experience.

3. **Search Bar:**
   - Integrated a search bar for efficient searching in both List and Grid views.

4. **Filter by Skills:**
   - Users can filter employees based on their skills for targeted results.

5. **Query Parameters:**
   - Added page, skill, and search values in query parameters to enhance URL tracking and bookmarking.

6. **Sort Functionality:**
   - Implemented sorting functionality for better data organization.

7. **Improved UI:**
   - Made enhancements to the user interface for a more polished and user-friendly design.

8. **Deletion of Employees:**
   - Users can now delete single or multiple employees for effective data management.

9. **Redux Integration:**
   - Configured the Redux store to manage the global state.
   - Implemented core functions for fetching data via Axios using Redux.
   - Set up actions and reducers to manage data flow and state for a more robust architecture.

These additions collectively contribute to a more powerful and user-centric Employee Listing Page, providing a rich set of features and improving the overall usability.

### CODE REVIEW

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -21,8 +21,7 @@ concurrency:
   cancel-in-progress: true
```
Comment (ashwin8448) [inline .github/workflows/deploy.yml:21]: Change branch name-"firebase integration"

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -21,8 +21,7 @@ concurrency:
...
 #Allow repo secrets
 env:
-  VITE_DATABASE_URL: ${{ secrets.VITE_DATABASE_URL }}
-  VITE_API_KEY: ${{ secrets.VITE_API_KEY }}
+  VITE_API_ENDPOINT: ${{ secrets.VITE_API_ENDPOINT }}
```
Comment (ashwin8448) [inline .github/workflows/deploy.yml:24]: Remove VITE_DATABASE_URL and VITE_API_KEY

File: src/components/Button/button.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    props.$isChildren ? `${colors.WHITE_COLOR}` : `transparent`};
+  cursor: pointer;
+  color: ${colors.SECONDARY_COLOR};
+  padding: ${(props) => (props.$isChildren ? `10px` : `0`)};
+  border-radius: 5px;
+  text-decoration: none;
+  gap: 5px;
+
+  // for the transition
+  backface-visibility: hidden; // hide the element's back face on rotation
```
Comment (ashwin8448) [inline src/components/Button/button.ts]: Is any rotation applied to buttons at any place?

File: src/components/Button/button.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  transition: ${(props) => !props.$noTransition && 'transform 0.2s'};
+
+  // disables text selection by the user
+  user-select: none;
+  -webkit-user-select: none;
+
+  // hint for the user agent to prioritize the processing of touch events
+  touch-action: manipulation;
+
+  //TODO: if a different color needs to be added for the icon if there is no text
```
Comment (ashwin8448) [inline src/components/Button/button.ts]: If completed, remove TODO.

File: src/components/DeleteModal/constants/constants.ts [hunk 5]
```diff
@@ -1,7 +1,9 @@
 export const DELETE_MODAL_HEADING = "Confirm user removal";
 
-export const CONFIRM_DELETE_TEXT = (employeeId: string) =>
-  `Are you sure you want to delete the employee ${employeeId}?`;
+export const CONFIRM_DELETE_TEXT = (employeeIdList: string[]) =>
+  `Are you sure you want to delete the employee${
+    employeeIdList.length > 1 ? "s" : ""
+  } ${employeeIdList.join(", ")} ?`;
```
Comment (ashwin8448) [inline src/components/DeleteModal/constants/constants.ts:6]: If possible display employee names also when deleting.

File: src/components/Loader/Loader.tsx [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import LoaderWrapper from "./loader.ts";
```
Comment (ashwin8448) [inline src/components/Loader/Loader.tsx:1]: There was an issue with the loader; it was not centered. We can work on it together.

File: src/core/config/constants.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+// TODO: Delete
+export enum SortDirection {
```
Comment (ashwin8448) [inline src/core/config/constants.ts:1]: Is this required?

File: src/core/interfaces/interface.ts [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+      skills: ISelectOptionProps[];
+    };
+  };
+}
+export interface IReceivingEmployeeData {
+  employees: IReceivingEmployee[];
+  count: number;
+  loading: boolean;
+}
+export interface IReceivingEmployee {
```
Comment (ashwin8448) [inline src/core/interfaces/interface.ts:23]: I had made some changes to the interfaces in formPage branch. We can discuss and implement the better one.

File: src/core/store/actions.ts [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+});
+export const setRoles = (roles: ISelectOptionProps[]) => ({
+  type: actionTypes.SET_ROLES,
+  payload: roles,
+});
+export const setSkills = (skills: ISelectOptionProps[]) => ({
+  type: actionTypes.SET_SKILLS,
+  payload: skills,
+});
+export const resetEmployeesGrid = () => {
```
Comment (ashwin8448) [inline src/core/store/actions.ts:52]: Why do we need this?

File: src/core/store/actionTypes.ts [hunk 10]
```diff
@@ -0,0 +1,13 @@
+export const SET_DATA_EMPLOYEES = "SET_DATA_EMPLOYEES";
+export const SET_EMPLOYEES = "SET_EMPLOYEES";
```
Comment (ashwin8448) [inline src/core/store/actionTypes.ts:2]: Difference between SET_EMPLOYEES and SET_EMPLOYESS_LIST/SET_EMPLOYEES_GRID?

File: src/core/store/actions.ts [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+    } catch (error) {
+      toast.error("No data is recieved", { toastId: "no-data" });
+      console.error("Error fetching data:", error);
+    } finally {
+      dispatch(setLoading(actionTypes.SET_LOADING, false));
+    }
+  };
+};
+
+export const fetchDropdownData = () => {
```
Comment (ashwin8448) [inline src/core/store/actions.ts:149]: I guess here the fetch happens sequentially one after the other. We will have to make these fetches simultaneously.

File: src/pages/routes.tsx [hunk 12]
```diff
@@ -1,17 +1,24 @@
...
+        {
+          path: "/error?statusCode=:statusCode",
+          element: <ErrorPage />,
         },
       ],
-      errorElement: <></>,
+      errorElement: <ErrorPage />,
     },
   ],
   { basename: import.meta.env.DEV ? "/" : "/HRM-App-React/" }
```
Comment (ashwin8448) [inline src/pages/routes.tsx]: Change repo name.

File: src/components/Footer/Footer.tsx
Comment (ashwin8448) [inline src/components/Footer/Footer.tsx:1]: Our header and footer stretches out when we zoom out. We will have to fix it.

File: src/core/store/reducers/filterReducer.ts
Comment (ashwin8448) [inline src/core/store/reducers/filterReducer.ts:1]: Is this file required?

**Review by ashwin8448 — CHANGES_REQUESTED** [review]
Added few comments. We can work on them together.

File: .github/workflows/deploy.yml (hunk 1 above)
Comment (seraGeorge) [reply .github/workflows/deploy.yml:21]: Resolved in a7625b4

File: .github/workflows/deploy.yml (hunk 2 above)
Comment (seraGeorge) [reply .github/workflows/deploy.yml:24]: Done

File: src/components/Button/button.ts (hunk 3 above)
Comment (seraGeorge) [reply src/components/Button/button.ts]: Resolved in 616328b0ae52390a300a1bc159a1971143c05f9d

File: src/core/store/actionTypes.ts (hunk 10 above)
Comment (seraGeorge) [reply src/core/store/actionTypes.ts:2]: SET_EMPLOYEES_LIST // for overriding the current employee list to enable pagination
SET_EMPLOYEES_GRID // for concatenating the current employee list with the new list to enable infinite loading

The reducers deals each action differently

File: src/components/Button/button.ts (hunk 4 above)
Comment (seraGeorge) [reply src/components/Button/button.ts]: Resolved in c1d862b1e940779d808ab55886138a9ced297f02

File: src/pages/routes.tsx (hunk 12 above)
Comment (seraGeorge) [reply src/pages/routes.tsx]: Resolved in c1d862b1e940779d808ab55886138a9ced297f02

File: src/core/store/reducers/filterReducer.ts
Comment (seraGeorge) [reply src/core/store/reducers/filterReducer.ts:1]: Resolved in c1d862b1e940779d808ab55886138a9ced297f02

File: src/core/config/constants.ts (hunk 7 above)
Comment (seraGeorge) [reply src/core/config/constants.ts:1]: enum will ensure that sort don't attain any other values other than the ones specified

File: src/components/Loader/Loader.tsx (hunk 6 above)
Comment (seraGeorge) [reply src/components/Loader/Loader.tsx:1]: done

File: src/core/interfaces/interface.ts (hunk 8 above)
Comment (seraGeorge) [reply src/core/interfaces/interface.ts:23]: done

File: src/core/store/actions.ts (hunk 9 above)
Comment (seraGeorge) [reply src/core/store/actions.ts:52]: reseting employees every time we move to the grid

File: src/core/store/actions.ts (hunk 11 above)
Comment (seraGeorge) [reply src/core/store/actions.ts:149]: changed

File: src/components/Footer/Footer.tsx
Comment (seraGeorge) [reply src/components/Footer/Footer.tsx:1]: done
